docker/jenkins/Jenkinsfile

```diff
@@ -1,46 +1,858 @@
+              unstashSuccessfulStage(env.SANITY_BNN_TEST_ZCU104, "xml_sanity_bnn_test_ZCU104")
+              unstashSuccessfulStage(env.SANITY_BNN_TEST_KV260_SOM, "xml_sanity_bnn_test_KV260_SOM")
+              unstashSuccessfulStage(env.END2END, "end2end")
+              unstashSuccessfulStage(env.BNN_BUILD_U250, "bnn_build_full_U250")
+              unstashSuccessfulStage(env.BNN_BUILD_PYNQZ1, "bnn_build_full_PynqZ1")
+              unstashSuccessfulStage(env.BNN_BUILD_ZCU104, "bnn_build_full_ZCU104")
+              unstashSuccessfulStage(env.BNN_BUILD_KV260_SOM, "bnn_build_full_KV260_SOM")
+              unstashSuccessfulStage(env.BNN_TEST_U250, "xml_bnn_test_U250")
+              unstashSuccessfulStage(env.BNN_TEST_PYNQZ1, "xml_bnn_test_PynqZ1")
+              unstashSuccessfulStage(env.BNN_TEST_ZCU104, "xml_bnn_test_ZCU104")
+              unstashSuccessfulStage(env.BNN_TEST_KV260_SOM, "xml_bnn_test_KV260_SOM")
+            }
+
+            // Combine individual HTML files to one single report
+            sh './run-docker.sh pytest_html_merger -i reports/ -o reports/test_report_final.html'
+
+            // Archive the XML & HTML test results
+            archiveArtifacts artifacts: "reports/*.xml"
+            archiveArtifacts artifacts: "reports/*.html"
+
+            archiveSuccessfulStage(env.SANITY_UT, "coverage_sanity_ut")
+            archiveSuccessfulStage(env.FPGADATAFLOW, "coverage_fpgadataflow")
+
+            // Plot what XML files were created during the test run
+            junit 'reports/*.xml'
+          }
+        }
+      }
+    }
+  }
+}
+
+void cleanPreviousBuildFiles(String buildDir) {
+  // Delete any build files from a previous build
+  // Previous build folders affect findCopyZip() and can cause the stage to fail
+  sh "rm -rf ${buildDir}/*"
+}
+
+void createMultiMarkerScript(String markers, String testResultsFilename, String additionalOptions) {
+  // Passing multiple markers when running ./run-docker.sh does not work with bash.
+  // Therefore, create a script to maintain the single quotes that surround the markers
+  sh """echo "#!/bin/bash
+python -m pytest -m \'${markers}\' --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html ${additionalOptions}" >> run-tests.sh
+    """
+
+  // Give permissions to script
+  sh 'chmod 777 run-tests.sh'
+}
+
+void runDockerPytestWithMarker(String marker, String testResultsFilename, String additionalOptions) {
+  sh """./run-docker.sh python -m pytest -m ${marker} --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html ${additionalOptions}"""
+}
+
+void findBoardBuildFiles(String board, String searchDir, String dirToFind) {
+  def result = sh(script: "find $searchDir -type d -name \"$dirToFind*\"", returnStdout: true).trim()
+  if (result.empty) {
+      error "Directory containing '$dirToFind' not found."
+  }
+	return result
```

Comment:
This method returns a non null value, I think it would be better to declare with "def" (https://www.tutorialspoint.com/groovy/groovy_methods.htm) or with the return type.

Reply:
Fixed